### DISCUSSION TITLE
Stacked pull requests/auto-rebase PRs

### DISCUSSION THREAD

**Discussion by alexec** [community/community#17880] [Pull Requests]
I often need to sequence changes and I want to create mulitple small PRs because they get reviewed in days, unlike big PRs can take weeks to get reviewed!

Those PRs are "stacked", i.e. I expect the first one to be approved and merged. The second should be rebased automatically.

**Comment by kiprasmel** [comment]
see https://github.com/kiprasmel/git-stacked-rebase

we don't have the ability to submit PRs or update the base branches yet, but will do soon:)

**Comment by anordal** [comment]
I like how Gerrit does it:

1. Each commit is perpetually identified by a never-changing [change-id](https://gerrit-review.googlesource.com/Documentation/user-changeid.html) (in the commit message, autogenerated by a commit hook).
2. You just push your whole stack as one branch to Gerrit. Gerrit presents each commit as a "change" that has to be reviewed, accepted & merged individually, but depends on its previous change in the changeset. Important feature: There is no button to mispresent the whole changeset as one (which is always the wrong way to review anything). You don't create pull requests either.
3. You can change each commit as much as you want and force-push your whole stack again. Gerrit sees what you did, beause the commit-ids stayed intact. If you force-pushed a trivial rebase, it says "trivial rebase", not "force push" – nice for your reviewers to know.

https://gerrit-review.googlesource.com/Documentation/intro-gerrit-walkthrough-github.html

**Comment by GirayTurker** [comment]
Had similar bug. **Switching between local repo's (ex: master to v=5.6.7) is locked, any commits and, editing commits are locked.**

The bug is visible like this on Git Bash:

<img width="637" alt="Screenshot 2023-12-16 135044" src="https://github.com/community/community/assets/89690556/71d4871b-004f-448c-9444-ab23ae5da8fa">

The pic shows the command to view ongoing rebase that it is not mapping any existing base (can be more than 1) along with last branch that system used to commit remote repo.

$ git checkout **your-repo-name**   will allow you to switch between branches. 
But it will not fix the main bug which is rebasing for last commit and, it will persisting to stay.

**$ git rebase --abort**  command will abort all rebase initiative(s). 
After hit the enter, it will abort all rebasing 

<img width="227" alt="Screenshot 2023-12-16 140442" src="https://github.com/community/community/assets/89690556/a352716e-22e0-4fe8-82a8-c434e07efe17">

Now, all mapping turns to normal, possible to observe it through your IDE.

**Reply by anordal** [reply]
Not related: This is a feature request about GitHub, while you are reporting a state superposition issue with Git. You have my sympathy (I keep saying that Git is unnecessarily and painfully stateful), but it's a different product.

**Comment by anordal** [comment]
See also #4477 for a proposal to add dependencies.

There are also merge bots that allow specifying dependencies for merge order:
https://docs.mergify.com/workflow/actions/merge/#pull-request-dependencies

I suppose this would be an instant hit, though Gerrit's approach is more automated.

**Comment by Skipants** [comment]
I would love to see first-party support for Stacked PRs. Using tools on top of Github for Stacked PRs hasn't made for the best experience.
